crates/sui-graphql-rpc/schema.graphql

```diff
@@ -0,0 +1,840 @@
+# - Pattern for upcasting and downcasting
+
+# Check Completeness Against Existing APIs
+# -------------
+# DONE Read
+# DONE - Objects
+# DONE - Transactions
+# DONE - Checkpoints
+# DONE - Events
+# DONE - ProtocolConfig
+# DONE - Chain Identifier
+# DONE Coin
+# DONE Extended
+# DONE Governance
+# DONE Indexer
+# DONE Move Utils
+# TODO Write
+
+# DONE Port designs from Notion
+# DONE - Display
+# DONE - SuiNS
+# DONE Data Consistency API
+# TODO Flesh out Connections
+# TODO - TotalTransactionBlocks for the PTB connection
+
+schema {
+  query: Query
+  subscription: Subscription
+}
+
+type Query {
+  # First four bytes of the network's genesis checkpoint digest
+  # (uniquely defines the network)
+  chainIdentifier: String!
+
+  # First and last checkpoint that the RPC can answer consistent
+  # queries for.
+  consistentRange: ConsistentRange
+
+  owner(address: SuiAddress!): Owner
+  object(address: SuiAddress!, version: U64): Object
+  address(address: SuiAddress!): Address
+
+  epoch(epochId: Int): Epoch
+  protocolConfig(protocolVersion: Int): ProtocolConfig
+
+  checkpoint(filter: CheckpointFilter): Checkpoint
+  checkpoints(order: Order!): CheckpointConnection
+
+  transactionBlock(digest: Digest!): TransactionBlock
+  transactionBlocks: TransactionBlockConnection
+
+  coinMetadata(coinType: TypeInst!): CoinMetadata
+
+  queryObjects(query: ObjectQuery): ObjectQueryConnection
+  queryEvents(filter: EventFilter, order: Order!): EventQueryConnection
+  queryTransactionBlocks(
+    filter: TransactionBlockFilter,
+    order: Order!,
+  ): TransactionBlockQueryConnection
+
+  resolveNameServiceAddress(name: String!): SuiAddress
+
+  networkMetrics: NetworkMetrics
+  moveCallMetrics: MoveCallMetrics
+  checkpointAddressMetrics(checkpoint: U64): AddressMetric
+  allEpochAddressMetrics(order: Order!): AddressMetricConnection
+}
+
+type Subscription {
+  events(filter: EventFilter): Event
+  transactions(filter: TransactionBlockFilter): TransactionBlock
+}
+
+# String containing 32B hex-encoded address
+scalar SuiAddress
+
+# String containing 32B Base58-encoded BLAKE2b256 digest
+scalar Digest
+
+# String representation of a 64 bit unsigned integer
+scalar U64
+
+# String representation of a 128 bit signed integer
+scalar I128
+
+# String representation of a 128 bit unsigned integer
+scalar U128
+
+# String containing Base58-encoded binary data.
+scalar Bytes
+
+scalar JSON
+
+# Fully qualified concrete type instantiation, with a flat
+# representation:
+#
+# ([package]::[module]::)?[name](<[param], ...>)?.
+scalar TypeInst
+
+# Move type from a function (parameter or return type) or struct
+# (field) signature.  May contain type parameters, referred to by
+# index.
+scalar TypeSignature
+
+# Filter either by the digest, or the sequence number, or neither, to
+# get the latest checkpoint.
+input CheckpointFilter {
+  digest: Digest
+  sequenceNumber: U64
+}
+
+enum Order { ASC, DESC }
+
+input ObjectQuery {
+  # one of these
```

Comment:
Thoughts on this style of rich query?

Reply:
This looks more powerful than what we expose in our rpcs today so i do worry that the complexity can cause very long running queries

Reply:
Vote for simplification for now

Reply:
The primitives are a little more general than what what we currently offer for object queries, but I don't think it's more expressive in any meaningful way?  Here's what we have at the moment:

https://github.com/MystenLabs/sui/blob/b611e669c8e1a09d9d75ad1a7691a5cacb522e47/crates/sui-json-rpc-types/src/sui_object.rs#L1107-L1138

I picked this set to try and simplify the schema.  I'm a little bit nervous about us relying on the expressivity of the query language to limit how long-running the query will be -- we should have a separate mechanism for bounding that.

Reply:
Ended up simplifying this initially!